eland/ml/transformers/lightgbm.py

```diff
@@ -0,0 +1,202 @@
+#  Licensed to Elasticsearch B.V. under one or more contributor
+#  license agreements. See the NOTICE file distributed with
+#  this work for additional information regarding copyright
+#  ownership. Elasticsearch B.V. licenses this file to you under
+#  the Apache License, Version 2.0 (the "License"); you may
+#  not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from typing import Optional, List, Dict, Any, Type
+from .base import ModelTransformer
+from .._model_serializer import Ensemble, Tree, TreeNode
+from ..ml_model import MLModel
+from .._optional import import_optional_dependency
+
+import_optional_dependency("lightgbm", on_version="warn")
+
+from lightgbm import Booster, LGBMRegressor  # type: ignore
+
+
+def transform_decider(decider: str) -> str:
+    if decider == "<=":
+        return "lte"
+    if decider == "<":
+        return "lt"
+    if decider == ">":
+        return "gt"
+    if decider == ">=":
+        return "gte"
+    raise ValueError(
+        "Unsupported splitting decider: %s. Only <=, " "<, >=, and > are allowed."
+    )
+
+
+class Counter:
+    def __init__(self, start: int = 0):
+        self._value = start
+
+    def inc(self) -> int:
+        self._value += 1
+        return self._value
+
+    def value(self) -> int:
+        return self._value
+
+
+class LGBMForestTransformer(ModelTransformer):
+    """
+    Base class for transforming LightGBM models into ensemble models supported by Elasticsearch
+
+    warning: do not use directly. Use a derived classes instead
+    """
+
+    def __init__(
+        self,
+        model: Booster,
+        feature_names: List[str],
+        classification_labels: Optional[List[str]] = None,
+        classification_weights: Optional[List[float]] = None,
+    ):
+        super().__init__(
+            model, feature_names, classification_labels, classification_weights
+        )
+        self._node_decision_type = "lte"
+        self._objective = model.params["objective"]
+
+    def build_tree(self, tree_json_obj: Dict[str, Any]) -> Tree:
+        tree_nodes = list()
+        next_id = Counter()
+        py_id_to_node_id: Dict[int, int] = dict()
+
+        def add_tree_node(tree_node_json_obj: Dict[str, Any], counter: Counter) -> None:
+            curr_id = py_id_to_node_id[id(tree_node_json_obj)]
```

Comment:
I'll just note that the use of `id()` odd, is there no other way to identify tree nodes? That's unfortunate :(

Reply:
There probably is, but I did not want to bother with it any longer. 

The underlying numbering scheme that lightgbm uses does not follow our convention. 

I could maybe fiddle with it more to effectively build the node list backwards with recursion, but this way seemed simple enough.

Reply:
I went ahead and used the `Counter` to mutate the node ids and recursively added nodes backwards (depth first). This was simple enough because I was already sorting the node list at the end :)